_2024/shell-advanced.md

```diff
@@ -0,0 +1,289 @@
+This is due to `>` by default only redirecting STDOUT. 
+
+The default file descriptors and their data streams are labelled as such:
+
+| File descriptor | Name   | Standard stream |
+|-----------------|--------|-----------------|
+| 0               | STDIN  | Program input   |
+| 1               | STDOUT | Program output  |
+| 2               | STDERR | Program errors  |
+
+Standard streams 1 and 2 can be redirected by specifying their ID before `>`, for example `cat /non/existent/file 2>error.txt`, which tries to output a non-existent file, and saves the error output in a file called `error.txt`. STDIN (fd 0) isn't often redirected, so you'll mostly be redirecting STDOUT and STDERR.
+
+If you have output for a file that you don't want to save or read, then you can redirect the standard stream to the device file `/dev/null`, which deletes any data sent to it.
+As an example, run `echo test` then `echo test >/dev/null` and notice that the error isn't logged in the latter example.
+This can also be used with STDERR, for example with `cat /non/existent/file 2>/dev/null`.
+
+Using this, data streams can be split, for example with `python3 -c "print('stdout goes here'); raise Exception('stderr goes here')" >stdout 2>stderr`.
+The command might look complicated, but all you need to know is the the `python3` command outputs data to STDOUT and STDERR.
+`>stdout.txt` redirects the STDOUT of the program to a file called `stdout.txt`, and `2>stderr` redirects the STDERR of the program to a file called `stderr.txt`.
+
+**Note:** the syntax `&>` can be used to redirect both STDOUT and STDERR, for example using `python3 -c "print('stdout goes here'); raise Exception('stderr goes here')" &>output.txt` to redirect all data to `output.txt`.
+
+You might have noticed that piping to `grep` doesn't filter error messages. This is because `grep` only filters STDOUT. This means that error messages that get sent through STDERR don't get processed by `grep`, and we need to redirect our standard streams to fix this.
+
+If we move the output from STDERR to STDOUT, `grep` will be able to filter it.
+We can do this with the syntax `2>&1`. This moves data from STDERR (fd `2`) to STDOUT (fd `1`), using an `&` before `1` to avoid ambiguity between writing to a file named `1`.
+
+Let's practice this by running `python3 -c "raise Exception('This goes to STDERR')"`. This runs the python code `raise Exception('This goes to STDERR')`, which throws an error and outputs context data around it. Trying to filter the output with `python3 -c "raise Exception('This goes to STDERR')" | grep STDERR` doesn't work, since the python error output goes to STDERR, but running `python3 -c "raise Exception('This goes to STDERR')" 2>&1 | grep STDERR` does, since it moves the STDERR stream to STDOUT, which `grep` does filter. As well as this, STDOUT can be redirected to STDERR to properly log errors, such as with `echo 'error here!' >&2`.
```

Comment:
Same as my other comment in `shell-intermediate.md` about breaking up large paragraphs with lots of code with examples of output.